File: matmult_dec.m
function p=matmult_dec(Adec, x, p)
% calcula p=p+A*x, A=diag(Adec(:,1))+Adec(:, 2)*Adec(:, 3)'

%int: i, j, k, N, K
%matrix: Adec, x, p
%double: s

N=size(x, 1);
K=size(x, 2);

if ( N== 1 )
    p=matmult(Adec, x, p);
else
for k=1:K
  % p=p+diag(Adec(:,1))*x
  for i=1:N
    p(i, k) = p(i, k)+Adec(i,1)*x(i, k);
  end
  
  % s=sum(x); p=p+Adec(:,2)*s
  s=0;
  for i=1:N
    s=s+x(i, k);
  end
  for i=1:N
    p(i, k)=p(i, k)+Adec(i, 2)*s;
  end
end
end

File: matmult_dec.cpp
#include "matmult_dec.h"
#include "update_utilities.h"

void matmult_dec(Matrix<double>& A, Matrix<double>& Qurmul, Matrix<double>& fh)
{
	int N = Qurmul.rows();
	int K = Qurmul.cols();

    Matrix<double> *ptQurmulk = new Matrix<double>(N, 1); Matrix<double> &Qurmulk = *ptQurmulk;

    if (N == 1)
        matmult(A, Qurmul, fh);
    else
    {
	    for (int k = 1; k <= K; k++)
        {
            /** p=p+diag(Adec(:,1))*x **/
            for (int i = 1; i <= N; i++) 
                fh(i, k) = fh(i, k) + A(i, 1)*Qurmul(i, k);        
            /** s=sum(x); p=p+Adec(:,2)*s **/
            get_col(Qurmulk, Qurmul, k);
            double s = sum1D(Qurmulk);
            for (int i = 1; i <= N; i++) 
                fh(i, k) = fh(i, k) + A(i, 2)*s;     
        }        
    }
    delete ptQurmulk;
}